### DISCUSSION TITLE
How to use jest.spyOn with @swc/jest?

### DISCUSSION THREAD

**Discussion by BrunoQuaresma** [swc-project/swc#7024] [Q&A] [answered]
I'm trying to use @swc/jest on my project but when I try it I get the following error:
```console
TypeError: Cannot redefine property: deleteTemplate
        at Function.defineProperty (<anonymous>)
```
```tsx
jest.spyOn(API, "deleteTemplate")
```
Thanks in advance!

**Comment by magic-akari** [comment, marked as answer]
Try this https://www.npmjs.com/package/jest_workaround

**Comment by adbutterfield** [comment]
I found a pattern that works for a project I'm migrating to swc at work. 
Sample here: https://github.com/adbutterfield/fast-refresh-express/blob/main/test/renderHelpers.test.ts

**Comment by flora8984461** [comment]
I am also stuck on this. The above solutions don't work for me because I am using a 3rd party library that uses swc, and I get this same error. My project does not use swc.  jest.mock() with mockReturnedValue and mockImplementation don't use my mock value.

Any suggestions are really appreciated.

**Comment by pedenys** [comment]
Found by chance a nice and easy workaround that does not involve changing anything in the non-test code (at least for me) : 

```ts
import * as externallibrary from "external-library";

jest.mock("external-library", () => ({
  __esModule: true,
  ...jest.requireActual("external-library"),
}));

describe("foo tests", () => {

  it("should bar if blop", () => {

    jest.spyOn(externallibrary, "externalMethod").mockReturnValue({
      isAuthenticated: true,
      isLoading: false,
    } as ReturnType<typeof externallibrary.externalMethod>);

   //...
  })
})
```

Thanks to iarroyo on [StackOverflow](https://stackoverflow.com/a/63374190)

**Comment by clintonb** [comment]
Historical context from https://github.com/swc-project/swc/issues/5059: `swc` "fixed" a bug that enables an invalid use of `jest.spyOn`.

`jest_workaround` is now [`swc_mut_cjs_exports`](https://www.npmjs.com/package/swc_mut_cjs_exports) and works well.

I used `jest_workaround` successfully for many months until a recent change in my repo broke it. I'm still not sure why, but I also realized that I only needed the workaround to resolve issues with a single spy on an exported function. I resolved this by converting the function a static method (which is better for our codebase in the long run), and removing `jest_workaround` altogether.
